File: leet_code/tries/replace_word.rb
class Trie

  def initialize
    @root = {}
  end

  def insert(word)
    node = @root
    word.each_char do |char|
      node[char] = {} if !node[char]
      node = node[char]
    end
    node[:prefix] = true
  end

  def shortest_prefix(word)
    node = @root
    replace = ""
    word.each_char do |char|
      return word if !node[char]
      replace += char
      node = node[char]
      return replace if node[:prefix]
    end
  end

end

def replace_words(dict, sentence)
  trie = Trie.new()
  dict.each {|prefix| trie.insert(prefix)}
  result = sentence.split(" ").map {|word| trie.shortest_prefix(word)}
  result.join(" ")
end

File: leet_code/tries/replace_word.js
class Trie {
    constructor() {
        this.root = {};
        this.prefix = false;
    }

    insert(word) {
        let node = this.root;
        for (let i = 0; i < word.length; i++) {
            if (!node[word[i]]) node[word[i]] = {};
            node = node[word[i]];
        }
        node.prefix = true;
    }

    shortestPrefix(word) {
        let node = this.root;
        let prefix = "";
        for (let i = 0; i < word.length; i++) {
            if (!node[word[i]]) return word;
            prefix += word[i];
            node = node[word[i]];
            if (node.prefix) return prefix;
        }
        return word;
    }
}


const replaceWords = (dict, sentence) => {
    const prefixTrie = new Trie(),
          arr = sentence.split(" ");

    for (let i = 0; i < dict.length; i++) {
        prefixTrie.insert(dict[i]);
    }

    for (let i = 0; i < arr.length; i++) {
        let replace = prefixTrie.shortestPrefix(arr[i]);
        if (replace !== arr[i]) arr[i] = replace;
    }
    return arr.join(" ");
};
